### ISSUE TITLE
vmselect: panics when storage has no free space for cache data

### ISSUE THREAD

**Issue by just1900** [VictoriaMetrics/VictoriaMetrics#4688]
### Describe the bug

I encountered an issue where vmselect panicked because the cache volume was full. The vmselect has a capacity of 2C/4GiB, and the cache volume size is 1GiB. 
I think the error is caused by a high volume of range queries that exhaust all available space, resulting in an unmarshal error.
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/6fd10e88713f0e181f4a488d3f89b57ccfd3701a/lib/storage/block_header.go#L143
I’m unsure about how to properly adjust the capacity of the cache volume size. Any suggestions?

### To Reproduce

mount a small size volume to vmselect and use it as cacheDataPath.

### Version

v1.82.1-cluster

### Logs

```
2023-07-09T08:40:31.622Z        panic        VictoriaMetrics/app/vmselect/netstorage/tmp_blocks_file.go:167        FATAL: unexpected non-empty tail left after unmarshaling data at offset 526770, size 137, tbfIdx 3; len(tail)=43
panic: FATAL: unexpected non-empty tail left after unmarshaling data at offset 526770, size 137, tbfIdx 3; len(tail)=43
```

and some other related logs:

```
cannot finalize temporary blocks files: cannot finalize temporary blocks file with 763 series: cannot write the remaining 94 bytes to "/select-cache/tmp/searchResults/3914280501": write /select-cache/tmp/searchResults/3914280501: no space left on device
```

### Screenshots

the volume usage
![image](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/16233178/ed597b31-e782-42af-be4d-62aa703fe67d)

### Additional information

It was observed during a benchmark, it seems that the resource usage are very tight during that time and vmselect is also experience an OOMKilled event.

**Comment by hagen1778** [comment]
Hello @just1900! 
vmselect uses the cache data path not for storing rollupCache, but also for temporary data during query processing when it can't fit avail memory.
I agree that vmselect could warn the user in a more graceful way by rejecting the query and logging an error. This can be done. But not is a simple fix, as it needs to start understanding errors returned by file system. This is also a transitive error, as multiple concurrent select requests can take arbitrary amount of disk space during selects processing.

As a workaround, I'd suggest allocating more disk space for vmselect so it has enough of space for persisting temporary data.

**Comment by just1900** [comment]
@hagen1778 Thanks for you reply, I made a rough estimation of how much space it may be required:
- Firstly, for the RollupResult Cache, it's `1/16 * allowed Memory`(seems like 1/8 on some other versions) https://github.com/VictoriaMetrics/VictoriaMetrics/blob/ad08d9c88483a888d589aab1f301ff1fab270413/app/vmselect/promql/rollup_result_cache.go#L88 this is obviously lower than the allocated memory for vmselect.
- Secondly, when considering the space used for temporary query data, it should not exceed `search.maxMemoryPerQuery * search.maxConcurrentRequests`, which again falls below the allocated memory for vmselect.

In conclusion, would it be appropriate to consider setting the capacity of select cache as twice the amount of memory allocated for vmselect?

**Comment by hagen1778** [comment]
The mentioned calculation are related to the limits of how much cache one query could occupy. If the size is exceeded, the result won't be cached. This supposed to protect cache pollution when one heavy query substitutes rollupCaches for many other queries.

The limit has nothing to do to the mentioned error, as the error comes from the OS. The vmselect limit can be set to arbitrary size, but if attached volume has no capacity - the process will fail on OS system call.

The suggested disk volume should be corresponding to the size of the received data parts during query processing on vmselect. While this one is not trivial to calculate, I'd suggest just allocating 10GB of disk volume and it is unlikely you'll ever face this problem again. Unless you configure vmselect to omit all query complexity limits and will try to `select *` (`{__name__!=""}`).

**Comment by just1900** [comment]
Thanks for clarifying. I misunderstood the maxMemoryPerQuery limit, as it is actually not related to the amount of temporary space used.

**Comment by divinesecret** [comment]
Lets suppose i select 6 months interval for a query and i have 100GB disk that is not enough to complete the query and show the results. So the only solution is to have more terabytes of disk space on `vmselect` node than the all storage nodes combined that holds that particular metric samples ?

**Comment by makasim** [comment]
https://github.com/VictoriaMetrics/VictoriaMetrics/pull/11428 fixes this scenario in which vmselect could panic:

- All data returned from A is successfully written to its temporary blocks file or is still buffered in memory. Therefore, the query does not fail, since partial responses are allowed.
- An intermediate flush of B's temporary blocks file fails, while Finalize() still succeeds for both shards — either because the remaining data for B is small enough to fit on disk, or because space was released by other queries in the meantime.

Then, when vmselect reads B's temporary blocks file, it can panic due to an unmarshal error.

The PR bubbles up any filesystem errors, which causes the client query to fail with a 503 status.

The bugfix is included in release [v1.151.0](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/victoriametrics/changelog/CHANGELOG.md#v11510) and [v1.148.3](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/victoriametrics/changelog/CHANGELOG.md#v11483), [v1.136.17](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/docs/victoriametrics/changelog/CHANGELOG.md#v113617) LTS lines.
